contentreview-impl/tool/src/java/org/sakaiproject/contentreview/servlet/SubmissionCallbackServlet.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.json.JSONObject;
+
+import org.sakaiproject.assignment.api.AssignmentSubmissionEdit;
+import org.sakaiproject.assignment.cover.AssignmentService;
+import org.sakaiproject.component.cover.ComponentManager;
+import org.sakaiproject.content.cover.ContentHostingService;
+import org.sakaiproject.content.api.ContentResourceEdit;
+import org.sakaiproject.contentreview.model.ContentReviewItem;
+import org.sakaiproject.contentreview.service.ContentReviewService;
+import org.sakaiproject.entity.api.ResourcePropertiesEdit;
+import org.sakaiproject.event.cover.NotificationService;
+import org.sakaiproject.tool.api.Session;
+import org.sakaiproject.tool.cover.SessionManager;
+
+/** 
+ * This servlet will receive callbacks from TII. Then it will process the data
+ * related to the submissions and resubmission and store it.
+ */
+
+@SuppressWarnings("deprecation")
+public class SubmissionCallbackServlet extends HttpServlet {
+	
+	private static Log M_log = LogFactory.getLog(SubmissionCallbackServlet.class);
+	
+	private ContentReviewService contentReviewService =
+				(ContentReviewService)ComponentManager.get("org.sakaiproject.contentreview.service.ContentReviewService");
+				
+	@Override
+	public void init(ServletConfig config) throws ServletException {
+		M_log.debug("init SubmissionCallbackServlet");
+		super.init(config);
+	}
+	
+	protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+		M_log.debug("doGet SubmissionCallbackServlet");
+		doPost(request, response);
+	}
+	
+	@SuppressWarnings("unchecked")
+	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+		M_log.debug("doPost SubmissionCallbackServlet");
+		String contentType = request.getContentType();
+		if ( contentType != null && contentType.startsWith("application/json") ) {
+			doPostJSON(request, response);
+		} else {
+			M_log.warn("SubmissionCallbackServlet received a not json call.");
+			response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+		}
+	}
+	
+	@SuppressWarnings("unchecked")
+    protected void doPostJSON(HttpServletRequest request, HttpServletResponse response) 
+            throws ServletException, IOException 
+    {
+		M_log.debug("doPostJSON SubmissionCallbackServlet");
+		StringBuilder sb = new StringBuilder();
+		BufferedReader reader = request.getReader();
+		try {
+			String line;
+			while ((line = reader.readLine()) != null) {
+				sb.append(line).append('\n');
+			}
+		} finally {
+			reader.close();
+		}
+		M_log.debug(sb.toString());
+		
+		JSONObject json = new JSONObject(sb.toString());
+		String submissionId = json.getString("lis_result_sourcedid");
+		int tiiPaperId = json.getInt("paperid");
+		//ext_outcomes_tool_placement_url parameter can also be processed if necessary
+		try{
+			Session session = SessionManager.getCurrentSession();
```

Comment:
We can probably add the ids for each object to the servlet urls, but that would make them unique but still accessible. If someone guesses the ids they will still be able to fake calls. Same with random numbers...